pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractReplicator.java

```diff
@@ -188,14 +222,34 @@ protected CompletableFuture<Boolean> isLocalTopicActive() {
         }, brokerService.executor());
     }
 
+    public synchronized CompletableFuture<Void> disconnect(boolean failIfHasBacklog) {
+        if (failIfHasBacklog && getNumberOfEntriesInBacklog() > 0) {
+            CompletableFuture<Void> disconnectFuture = new CompletableFuture<>();
+            disconnectFuture.completeExceptionally(new TopicBusyException("Cannot close a replicator with backlog"));
+            if (log.isDebugEnabled()) {
+                log.debug("[{}] Replicator disconnect failed since topic has backlog", replicatorId);
+            }
+            return disconnectFuture;
+        }
+
+        log.info("[{}] Disconnect replicator at position {} with backlog {}", replicatorId,
+                getReplicatorReadPosition(), getNumberOfEntriesInBacklog());
+        if (!tryChangeStatusToTerminating()) {
```

Comment:
It should be `tryChangeStatusToStopping()`?

And it's better add a log here.

Reply:
Added. And I moved `synchronized` after changing the status to `Stopping` successfully.